File: aml/client/types/media.ts
export type MediaType = 'book' | 'cd' | 'game';

export interface Media {
    _id: string;
    title: string;
    mediaType: MediaType;  // Restricted to specific values
    genre: string;
    releaseDate: Date;
    stock: number;
    description: string;
    imageUrl: string;
    borrowed: number;
    author?: string | null;  // Optional field
    publisher?: string | null;
    platform?: string | null;  // Optional, can be null
    artist?: string | null;  // Optional, can be null
    createdAt: Date;
}

export interface MediaSearchResponse {
    mediaItems: Media[],
    page: number,
    perPage: number,
    totalItems: number,
    totalPages: number
}

interface Genre {
    type: MediaType;
    genres: string[];
}

const genres: Genre[] = [
    {
        type: 'book',
        genres: [
            'Fiction',
            'Mystery',
            'Thriller',
            'Fantasy',
            'Science Fiction',
            'Romance',
            'Historical Fiction',
            'Horror',
            'Adventure',
            'Drama',
            'Non-fiction',
            'Biography',
            'Self-help',
            'True Crime',
            'Psychology',
            'Philosophy',
            'Business',
            'Politics',
            'Science',
            'Travel',
            'Cooking',
            'Health',
            'Religion',
            'Poetry',
            'Graphic Novels',
            'Comics',
        ]
    },
    {
        type: 'cd',
        genres: [
            'Pop',
            'Rock',
            'Classical',
            'Jazz',
            'Hip-Hop',
            'R&B',
            'Country',
            'Blues',
            'Electronic',
            'Reggae',
            'Punk',
            'Soul',
            'Folk',
            'Indie',
            'Alternative',
            'Metal',
            'Dance',
            'Gospel',
            'Techno',
            'Latin',
            'Soundtrack',
        ]
    },
    {
        type: 'game',
        genres: [
            'Action',
            'Adventure',
            'Role-Playing',
            'Shooter',
            'Strategy',
            'Simulation',
            'Sports',
            'Racing',
            'Puzzle',
            'Fighting',
            'Horror',
            'Platformer',
            'Multiplayer',
            'Open-World',
            'Survival',
            'Stealth',
            'Rhythm',
            'MOBA',
            'Battle Royale',
            'Simulation',
            'Visual Novel',
        ]
    }
];

export const getGenresForType = (type: MediaType): string[] => {
    const genreData = genres.find((genre) => genre.type === type);
    return genreData ? genreData.genres : [];
}

export const getAllGenres = (): string[] => {
    const allGenresSet = new Set<string>();
    genres.forEach((genreCategory) => {
        genreCategory.genres.forEach((genre) => {
            allGenresSet.add(genre);
        });
    });
    return Array.from(allGenresSet).sort();
}

File: aml/server/media-service/models/media.go
package models

import "time"

type Media struct {
	ID            string     `json:"_id"`
	Title         string     `json:"title"`
	MediaType     string     `json:"mediaType"`
	Genre         string     `json:"genre"`
	ReleaseDate   time.Time  `json:"releaseDate"`
	Stock         int        `json:"stock"`
	Description   string     `json:"description"`
	ImageUrl      string     `json:"imageUrl"`
	Borrowed      int        `json:"borrowed"`
	Author        *string    `json:"author,omitempty"`     // Optional field
	Publisher     *string    `json:"publisher,omitempty"`
	Platform      *string    `json:"platform,omitempty"` // Optional, can be null
	Artist        *string    `json:"artist,omitempty"`   // Optional, can be null
	CreatedAt     time.Time  `json:"createdAt"`
	IsBorrowed    *bool      `json:"isBorrowed,omitempty"`
	BorrowingRecord *BorrowingRecord `json:"borrowingRecord,omitempty"`
}
